Geomap: Add Symbol Alignment Options

drew08t opened the pull request:
Add options for Geomap markers to change anchor position.

- [x] Clean up displacement function
- [x] Take into account rotation for X and square symbols

After:
![Sep-01-2023 18-21-01](https://github.com/grafana/grafana/assets/60050885/13b9e533-781e-4d78-8ec7-293ac55f4f51)
![image](https://github.com/grafana/grafana/assets/60050885/0c27e319-eb68-4de4-abee-c16ffb9cba1b)


Closes #43936

drew08t committed:
Geomap: Add Symbol Anchor Options

drew08t committed:
Merge branch 'main' into drew08t/geomap-icon-anchors

drew08t committed:
Update displacement for svg case

drew08t committed:
For square and x, account for 45 degree offset

drew08t committed:
Simplify displacement function

drew08t committed:
Remove unused todo

drew08t committed:
Update test defaults to include anchor

drew08t committed:
Add missing anchor default to tests

nmarrs committed:
Merge remote-tracking branch 'origin' into drew08t/geomap-icon-anchors

nmarrs reviewed (commented):
Overall looks good! Testing locally though I noticed the the "Left / Right" + "Top / Bottom" options seem mirrored / backwards from what a user might expect them to be. I'd expect top to move the symbol up vs down etc.


https://github.com/grafana/grafana/assets/22381771/bd639d39-29dc-4409-80e7-3decd4fe2e97

drew08t commented on public/app/plugins/panel/geomap/style/markers.ts:
```diff
@@ -233,7 +248,9 @@ const makers: SymbolMaker[] = [
-          rotation: (rotation * Math.PI) / 180 + Math.PI / 4,
+          angle: Math.PI / 4,
+          rotation: (rotation * Math.PI) / 180,
+          displacement: getDisplacement(cfg.symbolAnchor ?? defaultStyleConfig.symbolAnchor, radius),
```
Yeah, good call there. Originally they were being called differently until I found a good workaround for X, square, and SVG types.

drew08t replied:
On second thought, since these are within the individual makers, which contain the cfg, it doesn't make sense to consolidate the calls.

drew08t commented:
> Overall looks good! Testing locally though I noticed the the "Left / Right" + "Top / Bottom" options seem mirrored / backwards from what a user might expect them to be. I'd expect top to move the symbol up vs down etc.
> 
>  testing.out.locally.mov

Thanks. I guess it depends on how the user interprets "anchor" position. Perhaps there is some better verbage or UX to use here?

nmarrs commented:
> Thanks. I guess it depends on how the user interprets "anchor" position. Perhaps there is some better verbage or UX to use here?

Yeah I guess it depends on what we are basing the positioning on (relative to the symbol or the location lat / lon). In my mind it makes more sense to position relative to the center point (lat, lon) vs the symbol itself. What do you think?

drew08t commented:
> > Thanks. I guess it depends on how the user interprets "anchor" position. Perhaps there is some better verbage or UX to use here?
> 
> Yeah I guess it depends on what we are basing the positioning on (relative to the symbol or the location lat / lon). In my mind it makes more sense to position relative to the center point (lat, lon) vs the symbol itself. What do you think?

I could go either way as long as we make it clear.

drew08t committed:
Move displacement function into utils and add test

drew08t committed:
Simplify anchor position options UX

drew08t commented:
I went ahead and added some tests for the displacement function and simplified the icon anchor position UX - I think it does a good job of getting across what we are controlling, which is the position of the icon anchor.

<img width="327" alt="image" src="https://github.com/grafana/grafana/assets/60050885/39b57426-f076-47d9-b104-fe7a22da62f9">

nmarrs commented on public/app/plugins/panel/geomap/editor/StyleEditor.tsx:
```diff
@@ -200,24 +216,48 @@ export const StyleEditor = (props: Props) => {
+              }
+            />
+          </Field>
+          <Field label={'Anchor Position'}>
```
tiny nits: I think there might be a component that we can use to group these two together and add the label at that level vs. on just the y anchor field

or maybe explicetely add labels to both (i.e. `Anchor Y Position` etc)

Also not sure on the wording of "anchor" vs just using "symbol" (i.e. `Symbol Position`) this makes more sense to me as user is determining behavior of the symbol aspect of the marker 🤷‍♂️

nmarrs reviewed (approved):
A few small nits - with those addressed I think this is good to go 💯

drew08t committed:
Change verbage to alignment and swap direction

drew08t committed:
Include missing alignment rename

drew08t committed:
Update tests

nmarrs committed:
Merge remote-tracking branch 'origin' into drew08t/geomap-icon-anchors

nmarrs reviewed (approved):
LGTM
